### ISSUE TITLE
Error: tls: oversized record received with length 20527

### ISSUE THREAD

**Issue by jaykm** [unixpickle/kahoot-hack#9]
Getting this error.
Error: tls: oversized record received with length 20527

**Comment by theswiftuniverse** [comment]
I am as well

**Comment by DennisSnijder** [comment]
same

**Comment by theswiftuniverse** [comment]
Does anyone have a solution for this?

**Comment by sciaschi** [comment]
Ill take a look into it, gonna see if i can get this working. Seems like a fun side project

**Comment by fishfishfish2104** [comment]
any luck

**Comment by fishfishfish2104** [comment]
im getting this error too

**Comment by unixpickle** [comment]
I actually ran this on two different operating systems (OS X and Linux) and it worked on both machines. Might I ask what operating system people are using when they get this error? Perhaps it's an issue with Go's TLS library, or a bug with the websocket API on Windows.

**Comment by fishfishfish2104** [comment]
I'm on Linux on a chromebook

**Comment by theswiftuniverse** [comment]
I'm on OSX

**Comment by jaykm** [comment]
Windows and OSX here

**Comment by fishfishfish2104** [comment]
It could be a problem with go I agree

**Comment by unixpickle** [comment]
More questions to see if I can reproduce:

 * Does this happen if you use a random, incorrect game pin (e.g., 123123)
 * Does this happen when you use a **correct** game pin?
 * Does this happen in the new version that I pushed up yesterday?
 * What version of Go are you using? You can find this by running the command `go version`.

**Comment by parthsaxena** [comment]
I just got the latest version you pushed yesterday. My Go version is "go1.5.1 darwin/amd64". I tried running it and I still got the same error, "Error: tls: oversized record received with length 20527". Here is a picture of me running the script in terminal. 
![image](https://cloud.githubusercontent.com/assets/13742336/12072488/1874a7c6-b095-11e5-91e3-0009661220a6.png) 

This was a correct game pin. 

Thanks for helping us :)

**Comment by fishfishfish2104** [comment]
I think I might have found out the problem I think its because of crypto/tls not being in the right directory 
I might be completely wrong

**Comment by unixpickle** [comment]
@parthsaxena it does not look like you successfully downloaded the latest version. The latest version (i.e. the one I pushed yesterday) does not prompt you for input. Rather, it takes everything as arguments. For instance, you'd do `go run kahoot-flood/main.go 165762 test 1` for the input you gave above.

In order to pull the new version (and make sure that it works), use `git reset HEAD --hard && git pull`. If it still doesn't work, let me know if the same credentials work from [http://hackkahoot.xyz](http://hackkahoot.xyz). That is a server running on my Linux server on Go 1.5.1.

**Comment by unixpickle** [comment]
@fishfishfish2104 I doubt it. I would not expect the program to compile at all if it could not find crypto/tls.

**Comment by parthsaxena** [comment]
Hey, sorry about that I just downloaded the ZIP of your repository before. I pulled it and I have the latest version now. I still am getting the same error however. Here is a picture: 
![image](https://cloud.githubusercontent.com/assets/13742336/12073696/6ed101f4-b0e1-11e5-93d1-87d46a71203e.png)

Again, thanks for helping!

**Comment by sciaschi1** [comment]
It seems like its an issue with the length of the decryption when its connecting. Its returning a value greater than the max its supposed to allow.

As per tls:
 "No valid TLS record has a type of 0x80, however SSLv2 handshakes start with a uint16 length where the MSB is set and the first record is always < 256 bytes long. Therefore typ == 0x80 strongly suggests an SSLv2 client."

though I could be wrong

**Comment by msemple1111** [comment]
@sciaschi1 That looks like it describing a different error.  ```tls: unsupported SSLv2 handshake received```
https://golang.org/src/crypto/tls/conn.go

This error is on line 569.

```
   533	Again:
   534		if c.rawInput == nil {
   535			c.rawInput = c.in.newBlock()
   536		}
   537		b := c.rawInput
```

```
   569		if n > maxCiphertext {
   570			c.sendAlert(alertRecordOverflow)
   571			return c.in.setErrorLocked(fmt.Errorf("tls: oversized record received with length %d", n))
   572		}
```

**Comment by unixpickle** [comment]
tl;dr: if you have this bug, please run this program and post its output: [https://gist.github.com/unixpickle/c3a3b3079ca847db621a](https://gist.github.com/unixpickle/c3a3b3079ca847db621a).

The precise error&mdash;including the length 20527&mdash;is what you get when you try to use HTTPS with an HTTP server. I know this because, when I change the hostname from "kahoot.it:443" to "kahoot.it:80", I get `tls: oversized record received with length 20527`.

The question is why some individuals are getting HTTP responses for their WSS (secure WebSockets) requests. It could be that these people are behind restrictive firewalls, or that they have some kind of parental controls enabled on their computers. It could even be that Kahoot has blacklisted their IPs for some reason.

To diagnose the issue, I want to know what response kahoot&mdash;or some kind of firewall&mdash;is giving to cause the error. To figure this out, I created a [small program](https://gist.github.com/unixpickle/c3a3b3079ca847db621a) that dumps the output of the WebSocket handshake. To run this, do `go run main.go GAME_PIN`. Note that you need a valid game pin to run this properly. Once the program is done running (which will take up to 10 seconds), copy all the output and paste it here. It will be hexadecimal output for the most part, and might look something like this:

```
tls error: tls: oversized record received with length 20527
Dumping output. Please wait roughly five seconds...
485454502f312e3120333031204d6f766564205065726d616e656e746c790d0a5365727665723a206f70656e72657374792f312e372e372e310d0a446174653a205361742c203032204a616e20323031362031373a35323a323920474d540d0a436f6e74656e742d547970653a20746578742f68746d6c0d0a436f6e74656e742d4c656e6774683a203139300d0a436f6e6e656374696f6e3a206b6565702d616c6976650d0a4c6f636174696f6e3a2068747470733a2f2f6b61686f6f742e69742f636f6d6574640d0a0d0a3c68746d6c3e0d0a3c686561643e3c7469746c653e333031204d6f766564205065726d616e656e746c793c2f7469746c653e3c2f686561643e0d0a3c626f6479206267636f6c6f723d227768697465223e0d0a3c63656e7465723e3c68313e333031204d6f766564205065726d616e656e746c793c2f68313e3c2f63656e7465723e0d0a3c68723e3c63656e7465723e6f70656e72657374792f312e372e372e313c2f63656e7465723e0d0a3c2f626f64793e0d0a3c2f68746d6c3e0d0a
Dumping output. Please wait roughly five seconds...

```

**Comment by msemple1111** [comment]
@unixpickle 
I ran the program twice, one with a game running and one without; here is the output:
First without a game running:
```
msemple-iMac:kahoot_test Michael$ go run main.go 838779
Dumping output. Please wait roughly five seconds...
485454502f312e312031303120537769746368696e672050726f746f636f6c730d0a5365727665723a206f70656e72657374792f312e372e372e310d0a446174653a205361742c203032204a616e20323031362031383a34363a333720474d540d0a436f6e6e656374696f6e3a20757067726164650d0a557067726164653a20576562536f636b65740d0a5365632d576562536f636b65742d4163636570743a2048536d726330734d6c59556b41476d6d354f50704732486147576b3d0d0a0d0a
```

Now with a game running:
```
msemple-iMac:kahoot_test Michael$ go run main.go 877234
Dumping output. Please wait roughly five seconds...
485454502f312e3120323030204f4b0d0a5365727665723a206f70656e72657374792f312e372e372e310d0a446174653a205361742c203032204a616e20323031362031383a34383a353020474d540d0a436f6e74656e742d547970653a206170706c69636174696f6e2f6f637465742d73747265616d0d0a5472616e736665722d456e636f64696e673a206368756e6b65640d0a436f6e6e656374696f6e3a206b6565702d616c6976650d0a0d0a32360d0a4d697373696e67206b65792066726f6e74656e643a3837373233342e6b61686f6f742e69740a0d0a300d0a0d0a
```

Interestingly they yielded very different results, but probably because the ```no.mobitroll.session``` cookie wasn't added.
It would be interesting to update the test to add the session cookie.

**Comment by unixpickle** [comment]
@msemple1111 do you usually experience the bug? If you did, I'd expect my test program to output the infamous TLS error. From your test results, I'd guess you weren't having the bug.

What were you saying about the `no.mobitroll.session` cookie? Both my actual client and the test program add that cookie.

I do not think the problem would have to do with browsers, since the Go hacks run independently of a browser. However, I like the hypothesis that proxies are messing with us. It seems the most plausible at this point.

**Comment by msemple1111** [comment]
@unixpickle I do have the TLS error, but not during that test. I can run http://websocket.org/echo.html on my browser so my internet connection is fine. So maybe that proxy theory is slightly off, perhaps its just go or kahoot not handling web sockets correctly.

I didn't see you added  ```no.mobitroll.session```  cookie, my bad, by the second output I just assumed that is what happened.

Also, I don't know why I said browsers I meant terminal+go+proxy+internet configuration, well I guess it was cause I was using a browser at the time.

Maybe the error is because in the actual client, you send the first connect request but say you supported  ```Long-polling``` but you don't, so when kahoot try's to downgrade, this error happens. Who knows.

**Comment by unixpickle** [comment]
@msemple1111 ah, that is very strange. The test program and the actual kahoot client start their connection exactly [the same way](https://github.com/unixpickle/kahoot-hack/blob/9934d5c2089d52cb1342db1e4dee6bc6f60a8235/kahoot/conn.go#L38-L41). I figured the TLS error was arising when I established the TLS connection, but perhaps it is happening later on in the [NewConn](https://github.com/unixpickle/kahoot-hack/blob/9934d5c2089d52cb1342db1e4dee6bc6f60a8235/kahoot/conn.go#L37) function. In other words, it might be happening once I've handed my connection off to gorilla/websocket.

Later I will make another test that might help establish when in the handshake the error is occurring.

**Comment by msemple1111** [comment]
@unixpickle I think it is something to do with gorilla/websocket. As this error is tls based I'm assuming it is part of the websoket <---> tls certificate handshake or the kahoot-hack client not feeding web socket with the correct input and then the error not being caught early enough.

**Comment by msemple1111** [comment]
If your still having this bug - goto my version written in python [here](https://github.com/msemple1111/kahoot-hack). It has the same basic tools but has more features, expanded prompts and easier instructions; it also has been designed to work behind a proxy by not utilising web sockets - so this bug will not occur.

https://github.com/msemple1111/kahoot-hack 

I started it before @unixpickle started to fix this version of kahoot-hack again so no one accusing me of copying pls.


P.S. If your a script kiddy trying to troll his/her teacher - Delete ```/Windows/System32```, it works a treat.

**Comment by unixpickle** [comment]
The reason I did not see this bug is that I was not using the latest version of gorilla/websocket. Once I upgraded, I got the error immediately and fixed it within minutes.

**Comment by fishfishfish2104** [comment]
cool mine works now thanks :)
